### ISSUE TITLE
Can't build

### ISSUE THREAD

**Issue by drbasic** [pkesist/buildpal#1]
Hi! I use phyton 2.7.6, and got error listed  below. What phyton version I need?

e:\chromium\buildpal-master>python setup.py build_ext
running build_ext
Traceback (most recent call last):
  File "setup.py", line 211, in <module>
    'console_scripts': ['buildpal = **main**']
  File "e:\chromium\depot_tools\python276_bin\lib\distutils\core.py", line 152,
in setup
    dist.run_commands()
  File "e:\chromium\depot_tools\python276_bin\lib\distutils\dist.py", line 953,
in run_commands
    self.run_command(cmd)
  File "e:\chromium\depot_tools\python276_bin\lib\distutils\dist.py", line 970,
in run_command
    cmd_obj = self.get_command_obj(command)
  File "e:\chromium\depot_tools\python276_bin\lib\distutils\dist.py", line 846,
in get_command_obj
    cmd_obj = self.command_obj[command] = klass(self)
  File "e:\chromium\depot_tools\python276_bin\lib\site-packages\setuptools-1.3.2
-py2.7.egg\setuptools__init__.py", line 69, in **init**
  File "e:\chromium\depot_tools\python276_bin\lib\distutils\cmd.py", line 64, in
 **init**
    self.initialize_options()
  File "setup.py", line 19, in initialize_options
    super().initialize_options()
TypeError: super() takes at least 1 argument (0 given)

**Comment by jivancic** [comment]
You need to use the latest Python - 3.4. BuildPal uses the new Python asyncio module which appeared in 3.4. Besides, it was never meant to work with Python 2.

http://buildpal.readthedocs.org/en/latest/building.html

**Comment by jivancic** [comment]
To save you the hassle, I uploaded a standalone package to SourceForge

https://sourceforge.net/projects/buildpal/files

**Comment by drbasic** [comment]
The project for a distributed build for Visual Studio would be a great help to me in the building of the chromium project. Build of chromium is made through ninja, which launches cl.exe.
Yesterday I built your project using python 3.4.1 and Visual studio 13 update 3. I needed to make a few changes version 11.0 -> 12.0 and for the project ninja needed to change the path for download. Standalone packages after cx_freeze did not start with an error (didn't save them). Python program does not work remotely. It seems that for cl.exe not installed correctly hooks, and you receive error that the source file is not found.
Thanks for the binaries, but x86 version is not working, it seems not happening interception run cl.exe and the building is run locally only. x64 version server and manager does not start with an error 

```
c:\Program Files\BuildPal>buildpal.exe server 
Traceback (most recent call last):
  File "D:\Python\Python34_x64\lib\site-packages\cx_freeze-4.3.2-py3.4-win-amd64
.egg\cx_Freeze\initscripts\Console3.py", line 27, in <module>
  File "Python/buildpal/__main__.py", line 58, in <module>
  File "Python/buildpal/__main__.py", line 43, in main
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\site-packages\buildpal-0.1-py3.4-win-amd64.eg
g\buildpal\server\__main__.py", line 1, in <module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\site-packages\buildpal-0.1-py3.4-win-amd64.eg
g\buildpal\server\runner.py", line 1, in <module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\site-packages\buildpal-0.1-py3.4-win-amd64.eg
g\buildpal\common\__init__.py", line 2, in <module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\site-packages\buildpal-0.1-py3.4-win-amd64.eg
g\buildpal\common\message.py", line 1, in <module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\asyncio\__init__.py", line 22, in <module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\asyncio\futures.py", line 8, in <module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2212, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 321, in _call_
with_frames_removed
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\concurrent\futures\__init__.py", line 17, in
<module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\concurrent\futures\process.py", line 55, in <
module>
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2237, in _find
_and_load
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 2226, in _find
_and_load_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1191, in _load
_unlocked
  File "D:\Python\Python34_x64\lib\importlib\_bootstrap.py", line 1161, in _load
_backward_compatible
  File "D:\Python\Python34_x64\lib\multiprocessing\connection.py", line 31, in <
module>
ImportError: cannot import name 'WAIT_ABANDONED_0'
```

**Comment by jivancic** [comment]
As far as binaries are concerned - could you try the latest ones? I found several issues with the ones I initially uploaded, one of which might explain the above error message.

I did not understand what went wrong with the version you build manually. I'd be willing to attempt to build Chromium myself. Feel free to contact me by e-mail (juraj.ivancic at gmail dot com). I'd appreciate more details on failure.

**Comment by drbasic** [comment]
DId you upade binaries in http://sourceforge.net/projects/buildpal/files ?

I tried the build again. Launched cl.exe on the remote machine happend, but it was not successful due to the fact that the source file was not found. So correct me if I'm wrong. We runs cl.exe on a remote machine but give the source file name like it was on local. Hooks must map from local machine names to a files in the temporary directory. However, this does not happened.

```
E:\chromium\chrome\src>b
ninja: Entering directory `out/Debug'
[22/18084] CC obj\native_client\src\tr...idator\validators64.validator_init.obj
FAILED: ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Vi
sual Studio 12.0\VC\bin\amd64\cl.exe" /nologo /showIncludes /FC @obj\native_clie
nt\src\trusted\validator\validators64.validator_init.obj.rsp /c ..\..\native_cli
ent\src\trusted\validator\validator_init.c /Foobj\native_client\src\trusted\vali
dator\validators64.validator_init.obj /Fdobj\native_client\src\trusted\validator
\validators64.c.pdb
c1 : fatal error C1083: Cannot open source file: 'E:\chromium\chrome\src\native_
client\src\trusted\validator\validator_init.c': No such file or directory
[22/18084] ACTION Generating resources from app/generated_resources.grd
ninja: build stopped: subcommand failed.
```

**Comment by drbasic** [comment]
Debug output from remote server:

```
DEBUG:root:Running '['C:\\Users\\D14E~1\\AppData\\Local\\Temp\\BuildPal\\Compile
rs\\5dcd53c3532828103ae2a34a5cc26106\\cl.exe', '/FoC:\\Users\\D14E~1\\AppData\\L
ocal\\Temp\\BuildPal\\Temp\\tmphgfuytgr\\tmp4n6osluu\\tmp624vl62e.obj', '/c', '/
nologo', '/showIncludes', '/FC', '-D', 'V8_DEPRECATION_WARNINGS', '-D', 'BLINK_S
CALE_FILTERS_AT_RECORD_TIME', '-D', '_WIN32_WINNT=0x0602', '-D', 'WINVER=0x0602'
, '-D', 'WIN32', '-D', '_WINDOWS', '-D', 'NOMINMAX', '-D', 'PSAPI_VERSION=1', '-
D', '_CRT_RAND_S', '-D', 'CERT_CHAIN_PARA_HAS_EXTRA_FIELDS', '-D', 'WIN32_LEAN_A
ND_MEAN', '-D', '_ATL_NO_OPENGL', '-D', '_HAS_EXCEPTIONS=0', '-D', '_SECURE_ATL'
, '-D', 'CHROMIUM_BUILD', '-D', 'COMPONENT_BUILD', '-D', 'TOOLKIT_VIEWS=1', '-D'
, 'USE_AURA=1', '-D', 'USE_ASH=1', '-D', 'USE_DEFAULT_RENDER_THEME=1', '-D', 'US
E_LIBJPEG_TURBO=1', '-D', 'ENABLE_ONE_CLICK_SIGNIN', '-D', 'ENABLE_PRE_SYNC_BACK
UP', '-D', 'ENABLE_REMOTING=1', '-D', 'ENABLE_WEBRTC=1', '-D', 'ENABLE_PEPPER_CD
MS', '-D', 'ENABLE_CONFIGURATION_POLICY', '-D', 'ENABLE_NOTIFICATIONS', '-D', 'E
NABLE_HIDPI=1', '-D', 'NO_TCMALLOC', '-D', 'ENABLE_EGLIMAGE=1', '-D', '__STD_C',
 '-D', '_CRT_SECURE_NO_DEPRECATE', '-D', '_SCL_SECURE_NO_DEPRECATE', '-D', 'NTDD
I_VERSION=0x06020000', '-D', '_USING_V110_SDK71_', '-D', 'ENABLE_TASK_MANAGER=1'
, '-D', 'ENABLE_EXTENSIONS=1', '-D', 'ENABLE_PLUGIN_INSTALLATION=1', '-D', 'ENAB
LE_PLUGINS=1', '-D', 'ENABLE_SESSION_SERVICE=1', '-D', 'ENABLE_THEMES=1', '-D',
'ENABLE_AUTOFILL_DIALOG=1', '-D', 'ENABLE_BACKGROUND=1', '-D', 'ENABLE_GOOGLE_NO
W=1', '-D', 'CLD_VERSION=2', '-D', 'CLD2_DATA_SOURCE=static', '-D', 'CLD_DATA_FR
OM_STATIC', '-D', 'ENABLE_FULL_PRINTING=1', '-D', 'ENABLE_PRINTING=1', '-D', 'WI
N_PDF_METAFILE_FOR_PRINTING=1', '-D', 'ENABLE_SPELLCHECK=1', '-D', 'ENABLE_CAPTI
VE_PORTAL_DETECTION=1', '-D', 'ENABLE_APP_LIST=1', '-D', 'ENABLE_SETTINGS_APP=1'
, '-D', 'ENABLE_MANAGED_USERS=1', '-D', 'ENABLE_MDNS=1', '-D', 'ENABLE_SERVICE_D
ISCOVERY=1', '-D', 'ENABLE_WIFI_BOOTSTRAPPING=1', '-D', '_CRT_NONSTDC_NO_WARNING
S', '-D', '_CRT_NONSTDC_NO_DEPRECATE', '-D', 'DYNAMIC_ANNOTATIONS_ENABLED=1', '-
D', 'WTF_USE_DYNAMIC_ANNOTATIONS=1', '-D', '_UNICODE', '-D', 'UNICODE', '-D', '_
DEBUG', '/wd4127', '/wd4351', '/wd4355', '/wd4503', '/wd4611', '/wd4100', '/wd41
21', '/wd4189', '/wd4244', '/wd4481', '/wd4505', '/wd4510', '/wd4512', '/wd4610'
, '/wd4800', '/wd4251', '/Od', '/W3', '/WX', '/Z7', '/GR-', '/Gy', '/GS', '/RTC1
', '/MDd', '/FS', '/Fdobj\\base\\third_party\\dynamic_annotations\\dynamic_annot
ations_win64.c.pdb', '/IC:\\Users\\D14E~1\\AppData\\Local\\Temp\\BuildPal\\Temp\
\tmphgfuytgr\\tmp4n6osluu', '/IC:\\Program Files (x86)\\Windows Kits\\8.0\\Inclu
de\\shared', '/IE:\\chromium\\chrome\\src', '/IC:\\Program Files (x86)\\Windows
Kits\\8.0\\Include\\um', '/IC:\\Program Files (x86)\\Microsoft Visual Studio 12.
0\\VC\\INCLUDE', 'E:\\chromium\\chrome\\src\\out\\Debug\\..\\..\\base\\third_par
ty\\dynamic_annotations\\dynamic_annotations.c']'
```

**Comment by jivancic** [comment]
BuildPal does not really change file paths when running cl.exe on the remote machine. The paths are the same as on local machine. But remote cl.exe itself is hooked so that windows APIs which try to open non-existing (client) files actually open substitutes. In short - remote machine gets a sort of a 'virtual filesystem'. This is needed so that debug information is generated correctly, like cl.exe was run on the client machine.

I have no idea why this file is missing - it shouldn't be. I am checking out chromium now,

**Comment by drbasic** [comment]
You set "virtual path" to cl.exe for includes, is it OK for debug info?

**Comment by jivancic** [comment]
Debug info is problematic, as PDB format is closed. BuildPal works around that by embedding debug information in the .obj file itself.

See http://buildpal.readthedocs.org/en/latest/bugs.html

**Comment by drbasic** [comment]
This is an example of the command you want to run remotely. 

```
"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\amd64_x86\cl.exe" /nologo /showIncludes /FC @obj\third_party\flac\src\libFLAC\libflac.float.obj.rsp /c ..\..\third_party\flac\src\libFLAC\float.c /Foobj\third_party\flac\src\libFLAC\libflac.float.obj /Fdobj\third_party\flac\libflac.c.pdb 
```

Note that there are two output files: obj and this particular PDB file. I apologize in this case, all file paths must be like on the local system. 
No need to perform linking remotely, only compiling needed.

**Comment by jivancic** [comment]
If you specify /Z7 instead of /Zi, .pdb file is not generated at all, and /Fd is redundant. Debug information gets stored in the .obj. In that case PDB is not needed at link time.

Not relevant to the above issue, but currently there is no support for remote linking.

**Comment by jivancic** [comment]
Id like to clarify further what really goes on here. Only source and header files are made 'virtual' on the server side. This is done because paths to these files become a part of debug information and, in order for debug info to make any sense on the client machine, these paths have to be client-local. 

Any compilation output file path need not be the same on the remote machine. Upon successful compilation, only the content is transferred to clients desired location. This is done for object files.

In a perfect world the same would be done for PDB files. However, PDB files are shared between different compiler invocations, and there is no way that I know of, to merge two PDB files. This is why /Zi -> /Z7 workaround is used.

To summarize - virtual paths are needed only for compiler input files.

**Comment by jivancic** [comment]
I have reproduced the issue.

The problem was introduced with the VS 2013 `cl.exe`. Previous versions allowed compiling a virtual file, say `C:\x\y\a.cpp`, even if `C:\x\y` directory did not exist. The new version bails out when `C:\x\y` is not found. To support this I'll need to update Win API hooks.

VS 2013 was just recently added and was not tested thoroughly. Thank you for the report.

I have added a test to detect this problem.

**Comment by drbasic** [comment]
Hi, Juraj!
I saw yesterday you made a small changes, they fix the bug?
Please let me know, you built chromium?

**Comment by jivancic** [comment]
Nope, for now I only added a test to catch this behavior. I am working on a fix, but it will take some time, as it involves a lot of trial-and-error. It is a bit of uncharted territory. When the fix is done I'll close this issue, so you'll be notified.

**Comment by jivancic** [comment]
Should be fixed now
